## **Beispiel von Data Analysis -Patentregister**

### Webseite vom DPMA : Suchwort "Software".
### Im Patentregister kann man alle Anmeldungen seit 2000 online finden.
#### Insgesamt  5.184 Ergebnisse mit dem wort "Software" im Titel.
#### Die Suchergebnisse werden als .xls Datei heruntergeladen.
#### Die .xls Datei wird ausgelesen und als Dataframe mit Python Pandas weiter verarbeitet:


In [506]:
import types
import pandas as pd
import numpy as np
from botocore.client import Config
import ibm_boto3

def __iter__(self): return 0

# Credentials IBM Cloud removed!
client_5f9bd298e5f7423d875763482f066f5f = ibm_boto3.client(service_name='s3',
    ibm_api_key_id='???',
    ibm_auth_endpoint="https://iam.eu-gb.bluemix.net/oidc/token",
    config=Config(signature_version='oauth'),
    endpoint_url='https://s3.eu-geo.objectstorage.service.networklayer.com')

body = client_5f9bd298e5f7423d875763482f066f5f.get_object(Bucket='pythonbasicsibmweek4-donotdelete-pr-pciwahsdx9xkbh',Key='treffer (1).xls')['Body']
# add missing __iter__ method, so pandas accepts body as file-like object
if not hasattr(body, "__iter__"): body.__iter__ = types.MethodType( __iter__, body )

df= pd.read_excel(body)
df.head(3)                 # apospasma toy pinaka me tis dyo mono protew seires


Suchanfrage: (TI = "software" or TIEZ = "software" or TIGZ = "software")  \
0                                                NaN                         
1                                       Aktenzeichen                         
2                                       100 00 673.6                         

        Unnamed: 1                Unnamed: 2  \
0              NaN                       NaN   
1  Schutzrechtsart                    Status   
2           Patent  Nicht anhängig/erloschen   

                                          Unnamed: 3            Unnamed: 4  \
0                                                NaN                   NaN   
1                                        Bezeichnung       IPC-Hauptklasse   
2  Software und technische Vorrichtung zur Herste...  G03B 37/00 (2000.01)   

           Unnamed: 5  Unnamed: 6                Unnamed: 7      Unnamed: 8  \
0                 NaN         NaN                       NaN             NaN   
1  IPC-Nebenklasse(n)  Anmeldetag  Erstveröffentlichungstag  Eintragungstag   
2                 NaN  11.01.2000                12.07.2001             NaN   

                                    Unnamed: 9            Unnamed: 10  \
0                                          NaN                    NaN   
1                             Anmelder/Inhaber              Erfinder    
2  brains3 GmbH & Co. KG, 94469 Deggendorf, DE  Erfinder Nichtnennung   

  Unnamed: 11  
0         NaN  
1   Vertreter  
2         NaN

In [44]:
x = type(df[['Unnamed: 3']])
x

pandas.core.frame.DataFrame

### Ein Überblick aller Spalten:

In [507]:
df.iloc[1, [0,1,2,3,4,5,6,7,8,9,10,11]]

Suchanfrage: (TI = "software" or TIEZ = "software" or TIGZ = "software")                Aktenzeichen
Unnamed: 1                                                                           Schutzrechtsart
Unnamed: 2                                                                                    Status
Unnamed: 3                                                                               Bezeichnung
Unnamed: 4                                                                           IPC-Hauptklasse
Unnamed: 5                                                                        IPC-Nebenklasse(n)
Unnamed: 6                                                                                Anmeldetag
Unnamed: 7                                                                  Erstveröffentlichungstag
Unnamed: 8                                                                            Eintragungstag
Unnamed: 9                                                                          Anmelde

In [508]:
df.loc[1, 'Unnamed: 4']

'IPC-Hauptklasse'

### Die Spaltenetiketten werden umbenannt:

In [546]:
# adikathisto ola ta labels ton columns giati ta pragmatika labels briskodan stin seira (row) 2.

df.rename(columns={'Suchanfrage: (TI = "software" or TIEZ = "software" or TIGZ = "software")': 'Aktenzeichen','Unnamed: 1': 'Schutzrechtsart','Unnamed: 2': 'Status','Unnamed: 3': 'Bezeichnung','Unnamed: 4': 'IPC-Hauptklasse','Unnamed: 5': 'IPC-Nebenklasse(n)','Unnamed: 6': 'Anmeldetag','Unnamed: 7': 'Erstveröffentlichungstag', 'Unnamed: 8': 'Eintragungstag', 'Unnamed: 9': 'Anmelder/Inhaber', 'Unnamed: 10': 'Erfinder', 'Unnamed: 11': 'Vertreter'},inplace=True)

                   # epeidi prin eiha valei  ,inplace=True allaxe to core df, opote eno esvisa tis akolouthes functions, opos to df.drop([1,2]), aftes apothikeftikan kai exakolouthoun na ishyoun horis na prepei na xanavalo tin edoli

                      ## edo diagrapsame tis dyo protes seires (rows) giati i proti eihe ta values NaN kai i defteri eihe os values ta labbels ton columns
                      ###  df.shift()[1:]  afti i edoli metathetei tin arithmisi syn 1 pros ta kato, dialdi an xekinouse apo to 1 (to index) tora xekinaei apto 2. den eho vrei pos paei omos pros ta piso i arithmisi.
                       #### enallaktiki methodos gia alalgi tis arithmis se memonomena omos simeia: df.rename({2: 1, 3:2}, axis=0)  # axis=0 einai to idio me to axis='index', kathorizei diladi to function pou theloume na ektelespoume, se poion axona tha pragmatopoiithei. opou 0=rows, 1= columns
                      
                       #####import numpy as np # df.index = np.arange(1, len(df)+1)  # Habemus Papam!!!  afti einai i teleftaia edoli pou edosa oste na allaxei i arithmisi kai na xekinaei apto 1 adi gia to 3
                                         
                                   

df.index = np.arange(1, len(df)+1)
df





Aktenzeichen Schutzrechtsart                    Status  \
1          100 00 673.6          Patent  Nicht anhängig/erloschen   
2          100 00 964.6          Patent  Nicht anhängig/erloschen   
3          100 01 097.0          Patent  Nicht anhängig/erloschen   
4          100 03 333.4          Patent  Nicht anhängig/erloschen   
5          100 15 474.3          Patent  Nicht anhängig/erloschen   
6          100 18 342.5          Patent  Nicht anhängig/erloschen   
7          100 21 955.1          Patent  Nicht anhängig/erloschen   
8          100 22 336.2          Patent  Nicht anhängig/erloschen   
9          100 23 820.3          Patent  Nicht anhängig/erloschen   
10         100 28 500.7          Patent  Nicht anhängig/erloschen   
11         100 28 818.9          Patent  Nicht anhängig/erloschen   
12         100 30 988.7          Patent  Nicht anhängig/erloschen   
13         100 32 866.0          Patent  Nicht anhängig/erloschen   
14         100 33 384.2          Patent  Nicht anhängig/erloschen   
15         100 37 397.6          Patent  Nicht anhängig/erloschen   
16         100 39 420.5          Patent  Nicht anhängig/erloschen   
17         100 44 844.5          Patent  Nicht anhängig/erloschen   
18         100 47 712.7          Patent  Nicht anhängig/erloschen   
19         100 48 942.7          Patent         Anhängig/in Kraft   
20         100 51 458.8          Patent  Nicht anhängig/erloschen   
21         100 53 214.4          Patent  Nicht anhängig/erloschen   
22         100 53 952.1          Patent  Nicht anhängig/erloschen   
23         100 55 171.8          Patent  Nicht anhängig/erloschen   
24         100 63 249.1          Patent  Nicht anhängig/erloschen   
25         100 65 668.4          Patent  Nicht anhängig/erloschen   
26         100 83 300.4          Patent  Nicht anhängig/erloschen   
27         100 84 798.6          Patent  Nicht anhängig/erloschen   
28         101 00 458.3          Patent  Nicht anhängig/erloschen   
29         101 01 644.1          Patent  Nicht anhängig/erloschen   
30         101 05 153.0          Patent  Nicht anhängig/erloschen   
...                 ...             ...                       ...   
5155  PCT/US2019/046695          Patent         Anhängig/in Kraft   
5156  PCT/US2019/046818          Patent         Anhängig/in Kraft   
5157  PCT/US2019/047520          Patent         Anhängig/in Kraft   
5158  PCT/US2019/047855          Patent         Anhängig/in Kraft   
5159  PCT/US2019/048478          Patent         Anhängig/in Kraft   
5160  PCT/US2019/049055          Patent         Anhängig/in Kraft   
5161  PCT/US2019/049185          Patent         Anhängig/in Kraft   
5162  PCT/US2019/049405          Patent         Anhängig/in Kraft   
5163  PCT/US2019/051285          Patent         Anhängig/in Kraft   
5164  PCT/US2019/051723          Patent         Anhängig/in Kraft   
5165  PCT/US2019/053296          Patent         Anhängig/in Kraft   
5166  PCT/US2019/053357          Patent         Anhängig/in Kraft   
5167  PCT/US2019/053958          Patent         Anhängig/in Kraft   
5168  PCT/US2019/053978          Patent         Anhängig/in Kraft   
5169  PCT/US2019/054061          Patent         Anhängig/in Kraft   
5170  PCT/US2019/054333          Patent         Anhängig/in Kraft   
5171  PCT/US2019/054366          Patent         Anhängig/in Kraft   
5172  PCT/US2019/054584          Patent         Anhängig/in Kraft   
5173  PCT/US2019/054975          Patent         Anhängig/in Kraft   
5174  PCT/US2019/055020          Patent         Anhängig/in Kraft   
5175  PCT/US2019/055042          Patent         Anhängig/in Kraft   
5176  PCT/US2019/055205          Patent         Anhängig/in Kraft   
5177  PCT/US2019/056152          Patent         Anhängig/in Kraft   
5178  PCT/US2019/057457          Patent         Anhängig/in Kraft   
5179  PCT/US2019/057559          Patent         Anhängig/in Kraft   
5180  PCT/US2019/057720          Patent         Anhängig/in Kraft   
5181  PCT

### Wir erkunden die Gesamtzahl von Eintragungen für jede Spalte 

In [416]:
df.count()

Aktenzeichen                5184
Schutzrechtsart             5184
Status                      5184
Bezeichnung                 5184
IPC-Hauptklasse             5184
IPC-Nebenklasse(n)          2529
Anmeldetag                  5184
Erstveröffentlichungstag    5184
Eintragungstag                61
Anmelder/Inhaber            1687
Erfinder                    5120
Unnamed: 11                  755
dtype: int64

In [433]:
df['Status'].count()  # mas dinei to synoliko arithmo timon mias stilis

5184

##### **Schauen wir uns die Spalte "Status" genau an, da sie 2 Werte annimmt.**

In [547]:
df.get_value(10, 'Status') # paradeigma enos tyhaiou tetragvnakioy os pros ti stili Status

/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:1: FutureWarning: get_value is deprecated and will be removed in a future release. Please use .at[] or .iat[] accessors instead
  if __name__ == '__main__':


'Nicht anhängig/erloschen'

In [548]:
df['Status'].unique()

array(['Nicht anhängig/erloschen', 'Anhängig/in Kraft'], dtype=object)

**column "Status" :Nicht anhängig/erloschen και Anhängig/in Kraft. (pending/granted not pending/not granted)**

In [549]:
df['Status'].value_counts()            #akoma kalytera me afto to function pairnoume se group tis pithanes times kai ton arithmo tou kathe group.

Nicht anhängig/erloschen    3797
Anhängig/in Kraft           1387
Name: Status, dtype: int64

In [550]:
df['Status'].value_counts(normalize=True)

Nicht anhängig/erloschen    0.732446
Anhängig/in Kraft           0.267554
Name: Status, dtype: float64

***Total Software patents since 2000: 3797 not valid.***

**73% not valid , 26% valid**

In [551]:
df['Schutzrechtsart'].value_counts()  

Patent             5122
Gebrauchsmuster      62
Name: Schutzrechtsart, dtype: int64

**Gebrauchsmuster(utility models) werden enfernt.
Insgesamt 5122 Eintragungen von Patenten.**

In [552]:
df[df.Schutzrechtsart == 'Patent'].groupby('Schutzrechtsart')['Status'].value_counts()

Schutzrechtsart  Status                  
Patent           Nicht anhängig/erloschen    3755
                 Anhängig/in Kraft           1367
Name: Status, dtype: int64

**Ratio: 3755 non-valid registry records to  1367 granted and still valid**

**Wir erkunden noch welche Patenthauptklassen im Datensatz repräsentiert sind (IPC-Hauptklasse):**

In [553]:
df['IPC-Hauptklasse'].value_counts()  

G06F 9/445 (2006.01)     344
G06F 9/44 (2006.01)      329
G06F 11/36 (2006.01)     187
H04L 29/06 (2006.01)     132
H04L 12/24 (2006.01)     113
G06F 17/30 (2006.01)     111
G06F 9/445 (2018.01)      98
H04L 29/08 (2006.01)      85
G06F 9/44 (2018.01)       82
G06F 9/45 (2006.01)       63
G06F 11/00 (2006.01)      55
n.a. ()                   54
G06F 1/00 (2000.01)       54
G06F 21/12 (2013.01)      50
G06F 9/50 (2006.01)       50
G06F 21/00 (2013.01)      48
G06F 21/57 (2013.01)      46
G06F 21/56 (2013.01)      46
G06F 15/16 (2006.01)      44
G06F 21/00 (2006.01)      41
G06F 1/00 (2006.01)       40
G06F 8/65 (2018.01)       38
G06F 9/46 (2006.01)       37
G06F 17/00 (2006.01)      37
G06F 21/10 (2013.01)      36
G06F 9/455 (2006.01)      35
H04L 12/26 (2006.01)      32
G06F 12/14 (2006.01)      30
H04L 12/28 (2006.01)      30
G06F 19/00 (2006.01)      29
                        ... 
H04L 25/49 (2006.01)       1
H04L 27/34 (2006.01)       1
C12Q 1/00 (2006.01)        1
G06F 5/00 (200

**Beschränken wir unsere Abfrage auf die Klassen, die mit einer Frequenz von mehr als 10 Zählungen auftreten:**

In [554]:
df['IPC-Hauptklasse'].value_counts().loc[lambda x: x>10]

G06F 9/445 (2006.01)     344
G06F 9/44 (2006.01)      329
G06F 11/36 (2006.01)     187
H04L 29/06 (2006.01)     132
H04L 12/24 (2006.01)     113
G06F 17/30 (2006.01)     111
G06F 9/445 (2018.01)      98
H04L 29/08 (2006.01)      85
G06F 9/44 (2018.01)       82
G06F 9/45 (2006.01)       63
G06F 11/00 (2006.01)      55
n.a. ()                   54
G06F 1/00 (2000.01)       54
G06F 21/12 (2013.01)      50
G06F 9/50 (2006.01)       50
G06F 21/00 (2013.01)      48
G06F 21/57 (2013.01)      46
G06F 21/56 (2013.01)      46
G06F 15/16 (2006.01)      44
G06F 21/00 (2006.01)      41
G06F 1/00 (2006.01)       40
G06F 8/65 (2018.01)       38
G06F 9/46 (2006.01)       37
G06F 17/00 (2006.01)      37
G06F 21/10 (2013.01)      36
G06F 9/455 (2006.01)      35
H04L 12/26 (2006.01)      32
G06F 12/14 (2006.01)      30
H04L 12/28 (2006.01)      30
G06F 19/00 (2006.01)      29
                        ... 
G06F 7/00 (2006.01)       15
H04L 12/801 (2013.01)     15
G06F 11/28 (2006.01)      15
G06F 21/53 (20

**Softwarepatente sind in 88 verschiedene Klassen unterteilt, wir könnten sie weiter in die 2-3 Hauptkategorien filtern**